File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/transport/ResponseDecoder.java
package in.org.quicko.silo.client.transport;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.errors.InvalidResponseException;

/**
 * Decodes a successful response by its content type: {@code 204} and an empty
 * body to null, {@code application/json} to a parsed tree, everything else to a
 * refusal when JSON was promised.
 */
public final class ResponseDecoder {
  private ResponseDecoder() {}

  public static JsonNode decode(
      int status, String contentType, String rawBody, TransportRequest request, JsonCodec json, boolean expectJson) {
    if (status == 204 || rawBody == null || rawBody.isEmpty()) return null;

    String type = contentType == null ? "" : contentType;
    if (type.contains("application/json")) return json.tree(rawBody);

    if (expectJson) throw new InvalidResponseException(request.method(), request.path(), type);
    return null;
  }
}

File: packages/silo-client-dart/lib/src/transport/response_decoder.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../errors/invalid_response_exception.dart';
import 'transport_request.dart';

/// A successful body as JSON text, or null for `204` and an empty body. A
/// route that promises JSON and answers anything else is refused.
abstract final class ResponseDecoder {
  static String? jsonText(http.Response response, TransportRequest request) {
    if (response.statusCode == 204 || response.bodyBytes.isEmpty) return null;
    final contentType = response.headers['content-type'] ?? '';
    if (!contentType.contains('application/json')) {
      throw InvalidResponseException(request.method, request.path, contentType);
    }
    return utf8.decode(response.bodyBytes);
  }
}
